Add the product vulnerability process (#1198)
mattermost/mattermost-handbook: operations/security/product-security/product-vulnerability-process.md

* [CWE-1395: Dependency on Vulnerable Third-Party Component](https://cwe.mitre.org/data/definitions/1395.html)
* [CWE-74: Improper Neutralization of Special Elements in Output Used by a Downstream Component ('Injection')](https://cwe.mitre.org/data/definitions/74.html)
* [CWE-922: Insecure Storage of Sensitive Information](https://cwe.mitre.org/data/definitions/922.html)
* [CWE-319: Cleartext Transmission of Sensitive Information](https://cwe.mitre.org/data/definitions/319.html)
* [CWE-327: Use of a Broken or Risky Cryptographic Algorithm](https://cwe.mitre.org/data/definitions/327.html)
* [CWE-74: Improper Neutralization of Special Elements in Output Used by a Downstream Component ('Injection')](https://cwe.mitre.org/data/definitions/74.html)
* [CWE-918: Server-Side Request Forgery (SSRF)](https://cwe.mitre.org/data/definitions/918.html)

### Backport Policy
For Mattermost server, the following backport policy applies:

#### Vulnerabilities with Low Score:
- Upcoming Release
- Extended Support Release

#### Vulnerabilities with Medium to Critical Score:
- Upcoming Release
- Extended Support Release
- Last 3 Releases

**Notes**:

> * The Mattermost server policy also applies to vulnerabilities in bundled plugins within the respective version

Reviewer: Can we be more explicit on what "bundled plugins" means?

Author: How about adding the following as a 3rd bullet in the Notes section?

> Bundled plugins are the plugins that are pre-installed in the Mattermost server. The list of bundled plugins can be found in Mattermost server's [Makefile](https://github.com/mattermost/mattermost-server/blob/master/server/Makefile#L141). Any other plugin is considered standalone.

> * There is no backport policy for Desktop, Mobile and standalone plugins.

Reviewer: Again, let's define what a "standalone" plugin is here to be more clear.

## Process Steps

### Report
Vulnerability reports are received from internal or external parties via one of the methods outlined above. The Product Security Team must acknowledge to the reporter that it received the report within 48h on business days and will investigate whether it’s a valid issue.

After the report was received, a “Security Vulnerability” ticket must be created in the Jira ticket that contains the following information:

* Title
* Description
* Category [If possible]
* Reporter
* Source
* Component

The security engineer that acknowledges the initial report must create a Mattermost Playbook Run using the "Security Vulnerability Playbook" and follow the process steps documented to continue with the Triage step. The playbook links to the Jira ticket created above, ensures the right process steps are completed, and informs relevant parties.

If the vulnerability was published on a news site/blog/etc. and is considered a public leak the Head of Marketing and VP Legal should be notified and invited to the playbook run.

### Triage

> After a vulnerability report was acknowledged it must be confirmed whether it is a valid finding within 96h of receiving it on business days. The security engineer should determine:

Reviewer: ```suggestion
After a vulnerability report was acknowledged it must be confirmed whether it is a valid finding within 96 hours of receiving it on business days. The security engineer should determine:
```

* Is the report a duplicate of a previous or current report from another reporter?
* Does the report have a security impact?
* Can the report be reproduced?

* If neither of the above hold true, does the report have an impact on any production infrastructure or systems and thus should be reviewed nevertheless?

After confirming that the report is valid the Product Security Team member must move the report to the next stage, called “Remediation” and must fill in the following information in the ticket:

* Vulnerability Category
* Steps to Reproduce
* Recommended Mitigation
* Security Update Text
* CVSS Score
* CVSS Severity
* Final Severity
* Mattermost Team
* Playbook Run link
* HackerOne Link
* Affected On-Premises Versions [Text]
* Affects Cloud [Yes/No]

**Additional auto-populated fields**

> * A Mattermost-specific vulnerability ID is automatically assigned when a report is moved from “Triage” to “Fix Development”)

Reviewer: ```suggestion
* A Mattermost-specific vulnerability ID is automatically assigned when a report is moved from “Triage” to “Fix Development”.)
```

> * A Due Date field is automatically populated when when a report is moved from “Triage” to “Fix Development” based on the Final Severity and the SLAs

Reviewer: ```suggestion
* A Due Date field is automatically populated when when a report is moved from “Triage” to “Fix Development” based on the Final Severity and the SLAs.
```

> If the vulnerability is not confirmed valid or considered a duplicate, the security engineer must notify the reporter about the decision. If we don’t consider it a vulnerability, the justification for our decision should be included both in the response to the reporter and as a comment in the Product Vulnerability ticket . The ticket should afterwards be closed either as duplicate or as invalid.

Reviewer: ```suggestion
If the vulnerability is not confirmed valid or considered a duplicate, the security engineer must notify the reporter about the decision. If we don’t consider it a vulnerability, the justification for our decision should be included both in the response to the reporter and as a comment in the Product Vulnerability ticket. The ticket should afterwards be closed either as duplicate or as invalid.
```

> If the report is deemed a security improvement (but not vulnerability), a development ticket should be created and the security vulnerability ticket closed with the development ticket added as a reference. A report is considered a security improvement when it is not exploitable without the presence of any additional vulnerabilities or only increases the

Reviewer: ```suggestion
If the report is deemed a security improvement (but not vulnerability), a development ticket should be created and the security vulnerability ticket closed with the development ticket added as a reference. A report is considered a security improvement when it is not exploitable without the presence of any additional vulnerabilities or only increases the general security posture.
```

> general security posture.

Reviewer: ```suggestion

```

### Remediation

> The Product Security Engineer must create a development ticket and link it to the vulnerability ticket. The Lead Engineer (or Product Manager) of the feature team responsible for the vulnerability fix and the Release Manager should be invited in the vulnerability’s playbook run channel and be informed about the vulnerability and its severity. The Lead Engineer must acknowledge the report and assign the development ticket to an engineer. The Lead Engineer should work together with the Product Manager on the prioritization of the reported security vulnerability.

Reviewer: ```suggestion
The Product Security Engineer must create a development ticket and link it to the vulnerability ticket. The Lead Engineer and Product Manager of the feature team responsible for the vulnerability fix and the Release Manager should be invited in the vulnerability’s playbook run channel and be informed about the vulnerability and its severity. The Lead Engineer must acknowledge the report and assign the development ticket to an engineer. The Lead Engineer should work together with the Product Manager on the prioritization of the reported security vulnerability.
```

* The vulnerability has a low severity

If in doubt, refer to the Backport Policy section above. After the fix development has finished and has been reviewed by peers, security and QA, a new dot release must be prepared.

In parallel with the development of the fix by the Engineering team, the Product Security Engineer must invite in the playbook run channel the Security Operations team, if the vulnerability is of “High” or “Critical” severity or if cloud installations are at risk of exploitation.
If cloud installations are at risk of exploitation, the Security Operations team must look at possible no-code mitigation steps to prevent the vulnerability from being exploited in the cloud until a fix is deployed.
If the vulnerability is of “High” or “Critical” severity, the Security Operations team must investigate for any exploitation attempts with the guidance of Product Security. If exploitation is confirmed, a separate process will be triggered which is outside of the scope of this document. The Product Security Engineer may invite the Security Operations team to look for exploitation attempts for “Medium” severity vulnerabilities on a case by case manner.

### Release Preparation

> After reviews have been completed, the developer must notify the Cloud SRE team and the Release Manager that a new cloud release version can be cut. Once the cloud release patch with the fix has been rolled out in the cloud, the release manager must cut a new dot release for all versions that received the security fix and release it. In case there are any problems during the release, the Release DevOps team should be notified to fix potential issues.

Reviewer: ```suggestion
After reviews have been completed, the developer must notify the Cloud SRE team and the Release Manager that a new cloud release version can be cut. Once the cloud release patch with the fix has been rolled out in the cloud, the Release Manager must cut a new dot release for all versions that received the security fix and release it. In case there are any problems during the release, the Release DevOps team should be notified to fix potential issues.
```

After the new versions are released, the Release Manager must update the [hall of fame list](https://mattermost.com/security-vulnerability-report) based on the Reporter field from the Jira ticket.

### Post Mortem
The Product Security Engineer must conduct a post mortem using the provided template for any vulnerabilities with “High” or “Critical” severity. The Product Security Engineer and the Lead Engineer of the feature team responsible for the vulnerability fix should meet and discuss:

* Why did this vulnerability occur?
* How can we verify that the vulnerability doesn’t exist in other parts of the product?
* What actions do we need to take to prevent it from happening again?

> The outcome of the postmortem must be action items based on the answers to the questions above. The Product Security Engineer must then fill the retrospective of the playbook run based on the answers in the above questions and create backlog tickets with the agreed action items.

Reviewer: ```suggestion
The outcome of the post mortem must be action items based on the answers to the questions above. The Product Security Engineer must then fill the retrospective of the playbook run based on the answers in the above questions and create backlog tickets with the agreed action items.
```

The Release Manager must create a draft for the email newsletter. The Marketing team and Product Security Engineering Lead must review the draft and send it out to all subscribers of the Security Updates mailing list.

### Responsible Disclosure

> This section is applicable If the vulnerability affects publicly released software artifacts.

Reviewer 2: ```suggestion
This section is applicable if the vulnerability affects publicly released software artifacts.
```

Author: Thanks, updated!
